components/playlist/renderer/playlist_js_handler.cc

```diff
@@ -0,0 +1,107 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/playlist/renderer/playlist_js_handler.h"
+
+#include "base/functional/callback.h"
+#include "base/logging.h"
+#include "content/public/renderer/render_frame.h"
+#include "gin/converter.h"
+#include "gin/function_template.h"
+#include "third_party/blink/public/common/browser_interface_broker_proxy.h"
+#include "third_party/blink/public/web/blink.h"
+
+namespace {
+
+template <typename Signature>
+void BindFunctionToObject(v8::Isolate* isolate,
+                          v8::Local<v8::Object> javascript_object,
+                          const std::string& function_name,
+                          const base::RepeatingCallback<Signature>& callback) {
+  v8::Local<v8::Context> context = isolate->GetCurrentContext();
+  // Get the isolate associated with this object.
+  javascript_object
+      ->Set(context, gin::StringToSymbol(isolate, function_name),
+            gin::CreateFunctionTemplate(isolate, callback)
+                ->GetFunction(context)
+                .ToLocalChecked())
+      .Check();
+}
+
+}  // namespace
+
+namespace playlist {
+
+PlaylistJSHandler::PlaylistJSHandler(content::RenderFrame* render_frame)
+    : render_frame_(render_frame) {
+  EnsureConnectedToClient();
+}
+
+PlaylistJSHandler::~PlaylistJSHandler() {}
+
+void PlaylistJSHandler::AddWorkerObjectToFrame(v8::Local<v8::Context> context) {
+  v8::Isolate* isolate = blink::MainThreadIsolate();
+  v8::HandleScope handle_scope(isolate);
+  if (context.IsEmpty()) {
+    return;
+  }
+
+  CreateWorkerObject(isolate, context);
+}
+
+bool PlaylistJSHandler::EnsureConnectedToClient() {
+  if (!client_.is_bound()) {
+    render_frame_->GetBrowserInterfaceBroker()->GetInterface(
+        client_.BindNewPipeAndPassReceiver());
+    client_.set_disconnect_handler(
+        base::BindOnce(&PlaylistJSHandler::OnClientDisconnect,
+                       weak_ptr_factory_.GetWeakPtr()));
+  }
+
+  return client_.is_bound();
+}
+
+void PlaylistJSHandler::OnClientDisconnect() {
+  client_.reset();
+  EnsureConnectedToClient();
+}
+
+void PlaylistJSHandler::CreateWorkerObject(v8::Isolate* isolate,
+                                           v8::Local<v8::Context> context) {
+  DVLOG(2) << __FUNCTION__;
+  v8::Local<v8::Object> global = context->Global();
+  v8::Local<v8::Value> pl_worker;
+  if (!global->Get(context, gin::StringToV8(isolate, "pl_worker"))
```

Comment:
It is not ok to add things to the global context in the main world

Reply:
it's also not ok to add JS apis to the main world. Did this get a security review?

Reply:
I guess this is an isolated world. There was a mention of a button "on page" so I was looking to see what that button did, but apparently the download is triggered from browser UI

Reply:
> It is not ok to add things to the global context in the main world
> it's also not ok to add JS apis to the main world. Did this get a security review?

No. I thought we were supposed to have a final review before the Android release(enbaling the feature), but we're having one for desktop first.

We needed MutationObserver for a site which updates video src later on(I think it was TED.) So I mimicked what cosmetic filter js handler does. And this mutation observer will be attached only to the offscreen contents which mirrors an actual tab's url.

---

If this shouldn't be done in this way, what should it be like?